Update getting-started-with-automixedprecisionmkl.md (#5)
intel-ai-tce/ai-documents: AEM/getting-started-with-automixedprecisionmkl.md

## Overview

> [Mixed precision](https://www.tensorflow.org/guide/mixed_precision) is the use of both 16-bit and 32-bit floating-point types in a model during training and inference to make it run faster and use less memory. Official TensorFlow supports this feature from TensorFlow 2.12 with all Intel® optimizations enabled by default. Intel® 4th Gen Xeon processor codenamed Sapphie Rapids(SPR) is the first Intel® processor to support Advanced Matrix Extensions(AMX) instructions, which help to accelerate matrix heavy workflows such as machine learning alongside also having BFloat16 support. Previous hardwares can be used to test functionality, but SPR provides the best performance gains. 

Reviewer: ```suggestion
[Mixed precision](https://www.tensorflow.org/guide/mixed_precision) is the use of both 16-bit and 32-bit floating-point types in a model during training and inference to make it run faster and use less memory. Official TensorFlow supports this feature from TensorFlow 2.12 with all Intel® optimizations enabled by default. Intel® 4th Gen Xeon processor codenamed Sapphire Rapids  is the first Intel® processor to support Advanced Matrix Extensions(AMX) instructions, which help to accelerate matrix heavy workflows such as machine learning alongside also having BFloat16 support. Previous generations can be used to test functionality, but  the Sapphire Rapids provides the best performance gains. 
```

Using one of the methods described here will enable a model written in FP32 data type to operate in mixed FP32 and BFloat16 data type. To be precise, it scans the data-flow graph corresponding to the model, and looks for nodes in the graph (also called as operators) that can operate in BFloat16 type, and inserts FP32ToBFloat16 and vice-versa Cast nodes in the graph appropriately.

For example: Consider a simple neural network with a typical pattern of Conv2D -> BiasAdd -> Relu with the default TensorFlow datatype FP32. TensorFlow data-flow graph corresponding to this model looks like below.

![amp-mkl](/content/dam/develop/external/us/en/images/automixedprecisionmkl-1.png)

The data-flow graph after porting the model to BFloat16 type looks like below.

![amp-mkl2](/content/dam/develop/external/us/en/images/automixedprecisionmkl-2.png)

Notice that 2 operators, Conv2D+BiasAdd and ReLU in the graph are automatically converted to operate in BFloat16 type. Also note that appropriate Cast nodes are inserted in the graph to convert TensorFlow tensors from FP32 type to BFloat16 type and vice-versa.

> Here are the options to enable BF16 mixed precision in TensorFlow.

Reviewer: ```suggestion
Here are the different ways to enable BF16 mixed precision in TensorFlow.
```

```bash
onednn_verbose,exec,cpu,matmul,brg:avx512_core_**amx_bf16**,undef,src_bf16::blocked:ba:f0 wei_bf16::blocked:ab:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,1600x128:128x10:1600x10
```

> Please use Python 3.8 or above for the best performance on Sapphire Rapids Xeon Scalable Processors.

## II. Mixed Precision for TensorFlow Hub models 
### Introduction
This session describes how to enable the auto-mixed precision for [TensorFlow Hub](https://www.tensorflow.org/hub) models using the tf.config API. Enabling this API will automatically convert the pre-trained model to use the bfloat16 datatype for computation resulting in an increased training throughput on the latest Intel® Xeon® scalable processor. These instructions works for inference or fine tuning usecase.

### Steps to do Transfer Learning with Mixed Precision on a Saved Model
We can use [this example for Transfer Learning](https://www.tensorflow.org/tutorials/images/transfer_learning_with_hub). By applying below experimental option "auto_mixed_precision_onednn_bfloat16" via tf.config API, auto mixed precision with bfloat16 will be enabled on the saved model.

```python

> tf.config.optimizer.set_experimental_options({'auto_mixed_precision_onednn_bfloat16':True})

Reviewer: ```suggestion
import tensorflow as tf
tf.config.optimizer.set_experimental_options({'auto_mixed_precision_onednn_bfloat16':True})
```

```

### Steps to enable Bfloat16 Inference for a Saved Model
Here's a [TF HUB inference example](https://www.tensorflow.org/hub/tutorials/tf2_object_detection). Same config as earlier can be used. 
```python

> tf.config.optimizer.set_experimental_options({'auto_mixed_precision_onednn_bfloat16':True})

Reviewer: ```suggestion
import tensorflow as tf
tf.config.optimizer.set_experimental_options({'auto_mixed_precision_onednn_bfloat16':True})
```

```

### End-to-end example using Mixed Precision for Transfer Learning 
For end-to-end example please refer to [this Transfer Learning sample](https://github.com/oneapi-src/oneAPI-samples/tree/master/AI-and-Analytics/Features-and-Functionality/IntelTensorFlow_Enabling_Auto_Mixed_Precision_for_TransferLearning) which uses a headless ResNet50v1.5 pretrained model from [TensorFlow Hub](https://www.tensorflow.org/hub) with ImageNet dataset. It also provides examples for performance improvements.

> > Remember to use Sapphire Rapids (SPR) server. say eg: C3 instance on GCP.

Reviewer: ```suggestion
> Remember to use Sapphire Rapids  server. say eg: C3 instance on GCP.
```

Using one of the methods described here will convert a model written in FP32 data type to operate in BFloat16 data type. To be precise, it scans the data-flow graph corresponding to the model, and looks for nodes in the graph (also called as operators) that can operate in BFloat16 type, and inserts FP32ToBFloat16 and vice-versa Cast nodes in the graph appropriately.

For example: Consider a simple neural network with a typical pattern of Conv2D -> BiasAdd -> Relu with the default TensorFlow datatype FP32. TensorFlow* data-flow graph corresponding to this model looks like below.

![amp-mkl](/content/dam/develop/external/us/en/images/automixedprecisionmkl-1.png)

The data-flow graph after porting the model to BFloat16 type looks like below.

![amp-mkl2](/content/dam/develop/external/us/en/images/automixedprecisionmkl-2.png)

Notice that 2 operators, Conv2D+BiasAdd and ReLU in the graph are automatically converted to operate in BFloat16 type. Also note that appropriate Cast nodes are inserted in the graph to convert TensorFlow* tensors from FP32 type to BFloat16 type and vice-versa.

Here are the options to enable BF16 mixed precision in TensorFlow

1. Keras mixed precision API

> 2. Transfer learning usecase with Tensorflow Hub SavedModel

Reviewer: This is for transfer learning as well inference on hub or saved models right?

Author: yes, that's right.
Should i rephrase as "Mixed Precision for Tensorflow Hub SavedModel"? and then later mention in the description that "this works for SavedModel inference and transfer learning usecases for models from TF Hub" ?

Reviewer: Yes.

brgconv_bwd_w:avx512_core_amx_bf16    150781.885700
```

> Please use Python 3.8 or above for the best performance on Sapphire Rapids Xeon Scalable Processors.

> ## II. Enable Mixed Precision on a Saved Model for Transfer Learning

Reviewer: Same comment as before, this will work on any hub saved model.

2022-08-19 00:06:08.750799: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:2303] Running auto_mixed_precision_onednn_bfloat16 graph optimizer
2022-08-19 00:06:08.751058: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:2240] Converted 3/11 nodes to bfloat16 precision using 0 cast(s) to bfloat16 (excluding Const and Variable casts)
```
> _No allowlist ops found typically refers to the data session._
> One can also use ONEDNN_VERBOSE logs as discussed earlier to verify AMX BF16 instructions.

> ## IV. Tips for best performance on Intel® Xeon CPUs

Reviewer: This seems unelated to mixed precision, can we remove this section?

Author: I added this to section to just bring awareness. But yes, I will remove it.

Reviewer: You can link to any existing optimization document.

### Using [numactl](https://man7.org/linux/man-pages/man8/numactl.8.html)
One should use numactl to run your code to obtain best performance. 
> Following are some examples and one can tweak the numactl parameters to get best performance for their specific usecase.

#### 1. for best throughput usecase performance, try using the whole socket or as many cores as posible
```bash
numactl --localalloc --cpunodebind 0 python conv2D_bf16.py
```
#### 2. for best latency usecase performance, try using fewer cores
```bash
numactl --localalloc --physcpunodebind=0-3 python conv2D_bf16.py
```

> ## Appendix 1 : Steps to convert a Pretrained fp32 Model to BFloat16

Reviewer: Lets remove this section also. We don't want people to use this, they should dynamically set auto_mixed.

Author: yes, will remove it. Seems too much details.

```
Again, notice 2 operators are converted into BFloat16 type. The script also supports dumping the model into protobuf binary file (.pb) or text file (.pbtxt).

> ## Appendix 2 : Controlling Ops that will be converted to BFloat16 type Automatically

Reviewer: Instead of keeping this in an Appendix, move under the legacy section (as this working only with legacy path)

Author: ok, will move it back under Legacy. Legacy section was getting too big so I had removed it from there and moved it here. 
Update: well, it will work for TF Hub case also. TF hub essentially uses grappler AMP beneath
